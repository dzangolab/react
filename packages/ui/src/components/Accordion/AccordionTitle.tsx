import React from "react";

type Properties = {
  title: string;
  icon: string;
  index: number;
  handleClick: (eventKey: number) => void;
};

const TabTitle: React.FC<Properties> = ({
  title,
  index,
  icon,
  handleClick,
}) => {
  return (
    <button
      onClick={() => handleClick(index)}
      type="button"
      className="accordion-trigger"
    >
      <div className="accordion-title">
        <span>{title}</span>
        {icon ? (
          <img src={icon} className="accordion-icon" alt={title} />
        ) : null}
      </div>
    </button>
  );
};

export default TabTitle;
