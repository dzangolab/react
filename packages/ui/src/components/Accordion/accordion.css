.accordion-list {
  align-items: flex-start;
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.accordion-list>li {
  width: 100%;
}

.accordion {
  border-radius: 7px;
  border: 2px solid hsl(0deg 0% 52%);
  margin: 0;
  padding: 0;
}

.accordion:focus-within {
  border-color: #1c4786;
}

.accordion>*+* {
  border-top: 0;
}

.accordion.active>*+* {
  border-top: 1px solid hsl(0deg 0% 52%);
}

.accordion-trigger {
  background: none;
  color: hsl(0deg 0% 13%);
  display: block;
  font-size: 1rem;
  font-weight: normal;
  margin: 0;
  outline: none;
  padding: 1em 1.5em;
  position: relative;
  text-align: left;
  width: 100%;
}

.accordion-trigger:focus,
.accordion-trigger:hover {
  background: hsl(216deg 94% 94%);
}

.accordion-trigger:focus {
  outline: 4px solid transparent;
}

.accordion>*:first-child .accordion-trigger,
.accordion>*:first-child {
  border-radius: 5px 5px 0 0;
}

.accordion>*:last-child .accordion-trigger,
.accordion>*:last-child {
  border-radius: 0 0 5px 5px;
}

.accordion button {
  border-style: none;
  cursor: pointer;
}

.accordion button::-moz-focus-inner {
  border: 0;
}

.accordion-title {
  align-items: center;
  border-radius: 5px;
  border: transparent 2px solid;
  display: flex;
  flex-direction: row;
  font-size: 1rem;
  justify-content: space-between;
  outline: none;
  padding: 0.25em;
  pointer-events: none;
  width: 100%;
}

.accordion-icon {
  border: 0;
  height: 1.5rem;
  pointer-events: none;
  position: absolute;
  right: 1.5rem;
  top: 50%;
  transform: translateY(-55%) rotate(0deg);
  width: 1.5rem;
}

.accordion-trigger:focus .accordion-icon path,
.accordion-trigger:hover .accordion-icon path {
  fill: none;
}

.accordion.active .accordion-icon {
  transform: translateY(-55%) rotate(-180deg);
}

.accordion-panel {
  margin: 0;
  padding: 0 1.5em;
}

.accordion-panel.hidden {
  border: 0;
}
