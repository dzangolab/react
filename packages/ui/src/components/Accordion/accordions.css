.accordion-list {
  list-style-type: none;
  margin: 0;
  padding: 0;
  display: flex;
}

.accordion-list>li {
  border-radius: 7px;
  border: 2px solid hsl(0deg 0% 52%);
  margin: 0;
  padding: 0;
}

.accordion-list>li:focus-within {
  border-color: #1c4786;
}

.accordion-list>li>*+* {
  border-top: 0;
}

.accordion-list>li>button {
  background: none;
  border-style: none;
  color: hsl(0deg 0% 13%);
  cursor: pointer;
  display: block;
  font-size: 1rem;
  font-weight: normal;
  margin: 0;
  outline: none;
  padding: 1em 1.5em;
  text-align: left;
  width: 100%;
}

.accordion-list>li>button:focus,
.accordion-list>li>button:hover {
  background: hsl(216deg 94% 94%);
}

.accordion-list>li>button>div {
  align-items: center;
  border-radius: 5px;
  border: transparent 2px solid;
  display: flex;
  flex-direction: row;
  font-size: 1rem;
  justify-content: space-between;
  outline: none;
  padding: 0.25em;
  pointer-events: none;
  width: 100%;
}

.accordion-list>li>button>div>img {
  border: 0;
  height: 1.5rem;
  pointer-events: none;
  width: 1.5rem;
}

/* TODO: remove */

.accordion-list.horizontal>li>button>div>img {
  transform: rotate(270deg);
}

.accordion-list.horizontal>li.active>button>div>img {
  transform: rotate(90deg);
}

.accordion-list.vertical>li>button>div>img {
  transform: rotate(0deg);
}

.accordion-list.vertical>li.active>button>div>img {
  transform: rotate(-180deg);
}

/* ------------ */

.accordion-list.horizontal {
  flex-direction: row;
  align-items: stretch;
}

.accordion-list.horizontal>li {
  display: flex;
  flex-direction: row;
}

.accordion-list.horizontal>li>button {
  width: min-content;
  display: flex;
  justify-content: space-between;
  transform: rotate(-180deg);
}

.accordion-list.horizontal>li>button>div {
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  width: min-content;
}

.accordion-list.horizontal>li>button>div>img {
  margin-top: var(--accordion-title-gap, 2rem);
}

.accordion-list>li>button>div>span {
  font-size: var(--accordion-title-size, 1rem);
}

.accordion-list.horizontal>li>button>div>span {
  writing-mode: vertical-lr;
  text-orientation: sideways-right;
  width: min-content;
}

.accordion-list.vertical {
  align-items: flex-start;
  flex-direction: column;
}

.accordion-list.vertical>li {
  width: 100%;
}

.accordion-list.vertical>li>button>div>span {
  width: max-content;
}

.accordion-list.vertical>li.active>*+* {
  border-top: 1px solid hsl(0deg 0% 52%);
}

.accordion-list.vertical>li>*:first-child>button,
.accordion>*:first-child {
  border-radius: 5px 5px 0 0;
}

.accordion-list.vertical>li>*:last-child>button,
.accordion-list.vertical>li>*:last-child {
  border-radius: 0 0 5px 5px;
}

.accordion-list.vertical>li:not(.active)>button:focus,
.accordion-list.vertical>li:not(.active)>button:hover {
  border-radius: 5px;
}
