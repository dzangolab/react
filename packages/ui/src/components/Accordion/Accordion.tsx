import React, { useId, useState } from "react";

import Collapse from "../Collapse";
import AccordionTitle from "./AccordionTitle";
import "./accordion.css";

import type { ReactElement } from "react";

type Properties = {
  defaultActiveKey?: number;
  children: ReactElement[];
};

const Accordion: React.FC<Properties> = ({ defaultActiveKey, children }) => {
  const id = useId();
  const [active, setActive] = useState(defaultActiveKey);

  return (
    <ul className="accordion-list" data-testid="accordion-list">
      {children.map((item, index) => (
        <li
          className={`accordion ${active === index ? "active" : ""}`}
          key={`${id}-${index}`}
        >
          <AccordionTitle
            handleClick={setActive}
            icon={item.props.icon}
            index={index}
            title={item.props.title}
          />
          <Collapse isOpen={active === index}>{item}</Collapse>
        </li>
      ))}
    </ul>
  );
};

export default Accordion;
