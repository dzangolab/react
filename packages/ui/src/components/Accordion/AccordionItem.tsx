import "./accordion.css";

import React, { useContext } from "react";

import Collapse from "../Collapse";
import { ToggleContentContext } from "../ToggleContent/ToggleContentProvider";

import type { ReactNode } from "react";

interface Properties {
  children: ReactNode;
  eventKey: string;
  label: string;
}

const AccordionItem = ({ children, eventKey, label }: Properties) => {
  const appContext = useContext(ToggleContentContext);
  if (!appContext) return null;

  const { active, handleClick } = appContext;
  const open = eventKey === active;

  return (
    <div className="accordion">
      <button
        onClick={() => handleClick(eventKey)}
        type="button"
        aria-expanded={open}
        className="accordion-trigger"
        aria-controls={`${eventKey}-body`}
        id={eventKey}
      >
        <span className="accordion-title">
          {label}
          <span className="accordion-icon"></span>
        </span>
      </button>

      <div
        id={`${eventKey}-body`}
        role="region"
        aria-labelledby={eventKey}
        className={`accordion-panel ${!open ? "hidden" : ""}`}
      >
        <Collapse isOpen={open}>{children}</Collapse>
      </div>
    </div>
  );
};

export default AccordionItem;
