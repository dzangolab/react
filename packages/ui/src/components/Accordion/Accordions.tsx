import React, { ReactElement, ReactNode, useState } from "react";

import AccordionTitle from "./AccordionTitle";
import "./accordion.css";

type Properties = {
  icon?: ReactNode;
  children: ReactElement[];
};

const Accordions: React.FC<Properties> = ({ icon, children }) => {
  const [selected, setSelected] = useState(0);

  return (
    <ul className="accordion-list">
      {children.map((item, index) => (
        <li key={index}>
          <div className="accordion">
            <AccordionTitle
              title={item.props.title}
              eventKey={index}
              selected={selected}
              setSelected={setSelected}
              icon={icon}
            />
            {selected === index ? children[selected] : null}
          </div>
        </li>
      ))}
    </ul>
  );
};

export default Accordions;
