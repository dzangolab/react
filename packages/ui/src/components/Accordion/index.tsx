import React, { useState } from "react";

import Collapse from "../Collapse";

import type { ReactNode } from "react";
import "./accordion.css";

interface Properties {
  accordionIcon?: ReactNode;
  children: ReactNode;
  id: string;
  label: string;
  isOpen?: boolean;
}

const Accordion = ({
  accordionIcon,
  children,
  id,
  isOpen,
  label,
}: Properties) => {
  const [open, setOpen] = useState(isOpen || false);
  const toggle = () => setOpen(!open);

  return (
    <div className="accordion">
      <button
        onClick={toggle}
        type="button"
        aria-expanded={open}
        className="accordion-trigger"
        aria-controls={`${id}-body`}
        id={id}
      >
        <span className="accordion-title">
          {label}
          {accordionIcon || <span className="accordion-icon"></span>}
        </span>
      </button>

      <div
        id={`${id}-body`}
        role="region"
        aria-labelledby={id}
        className={`accordion-panel ${!open ? "hidden" : ""}`}
      >
        <Collapse isOpen={open}>{children}</Collapse>
      </div>
    </div>
  );
};

export default Accordion;
