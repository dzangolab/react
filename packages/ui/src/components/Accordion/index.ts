import AccordionItem from "./AccordionItem";

export { AccordionItem };
