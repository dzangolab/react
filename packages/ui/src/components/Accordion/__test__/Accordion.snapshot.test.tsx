import React from "react";
import renderer, { create } from "react-test-renderer";
import { expect, test } from "vitest";

import Accordions from "../Accordions";
import Accordion from "../Accordion";

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as renderer.ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const component = create(
    <Accordions>
      <Accordion title="Accordion one">
        <p>Some text</p>
      </Accordion>
      <Accordion title="Accordion two">
        <p>Some text</p>
      </Accordion>
    </Accordions>
  );
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
