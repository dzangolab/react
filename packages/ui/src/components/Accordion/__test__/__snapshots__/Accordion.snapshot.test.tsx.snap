// Vitest Snapshot v1

exports[`Component matches snapshot 1`] = `
<ul
  className="accordion vertical"
>
  <li
    className=""
  >
    <button
      onClick={[Function]}
      type="button"
    >
      <img
        alt="Topic 1"
        src="icon1.jpg"
      />
      <span>
        Topic 1
      </span>
      <img
        src="arrow-up.jpg"
      />
    </button>
  </li>
  <li
    className=""
  >
    <button
      onClick={[Function]}
      type="button"
    >
      <img
        alt="Topic 2"
        src="icon2.jpg"
      />
      <span>
        Topic 2
      </span>
      <img
        src="arrow-up.jpg"
      />
    </button>
  </li>
</ul>
`;
